streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorContextImpl.java

```diff
@@ -288,7 +301,31 @@ private <K, V> void forwardInternal(final ProcessorNode<K, V, ?, ?> child,
                                         final Record<K, V> record) {
         setCurrentNode(child);
 
-        child.process(record);
+        try {
```

Comment:
I am not sure if this is the correct location to check for the exception and call the exception handler. What if the exception is thrown in the first processor of the topology. The first processor gets its input records from the stream task, not from the processor context.
I think a good place would be `StreamTask#doprocess()`. Did you consider that?

Reply:
We had a lot of hesitation regarding where to put the try-catch clause.

The StreamTask#doProcess() method doesn't seem to allow us to identify the node that throws the exception; the node will always be the source node.

This is why we decided to place the try-catch clause in the ProcessorContextImpl#forwardInternal method

Of course, we are fully open to discussing that.

Reply:
As far as I can see, by putting the catching in `ProcessorContextImpl#forwardInternal ` we miss exceptions that are thrown by the first processor. That is not correct. We need to find a better place.

Reply:
@cadonna Catching exceptions right here: https://github.com/apache/kafka/blob/0f0c9ecbf330923ad653cc2ff4fca6c4dced1cf7/streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java#L847 was our very first intention before we decided to move it to `ProcessorContextImpl#forwardInternal`.

Unless we're wrong, we cannot get the precise node name where the exception occured at `StreamTask#doProcess` level.

Could we:
- catch exceptions at `StreamTask#doProcess` level
- catch exceptions at `ProcessorContextImpl#forwardInternal` level, but rather than rethrowing a `StreamsException` (https://github.com/loicgreffier/kafka/blob/960c2a3153b30e48963387b5756b0310275bf48b/streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorContextImpl.java#L320), it can throw a `new ChildNodeProcessingException(child.name, e)` caught and handled at StreamTask#doProcess level. 

`ChildNodeProcessingException` would be a new exception acting as a wrapper of the root cause, that additionally provides the name of the child node in which an exception occurs... letting the first node know about something happened in one of its children.

Hope it is clear, if it sounds like a plan to you, we can give a try to this impl

Reply:
@loicgreffier I came up with one option how to add the processing exception handler to the processor node in the following draft PR: https://github.com/apache/kafka/pull/16137

This will create a exception handler per task. BTW, the KIP does not say anything about how many instances of the exception handler are created. Is it one per processor, one per task, one per thread, or even one per Streams client? The last one would require a thread-safe implementation of the handler. A user needs to know the granularity of the exception handler when implementing a handler that maintains state, e.g., counting the exceptions. I am in favor of a handler instance per task.

Let me know what you think.

Reply:
@cadonna 

PR has been updated from https://github.com/apache/kafka/pull/16137.

We gave a try to register the `processingExceptionHandler` in `InternalTopologyBuilder#rewriteTopology`. 
- The streams config is available already. It does not impact either `InternalTopologyBuilder#buildTopology` or `InternalTopologyBuilder#buildSubtopology`, and it deals with Topology Test Driver setup as well